rfcs/0062-content-addressed-paths.md

```diff
@@ -0,0 +1,286 @@
+    inputs by their real output path.
+    In that case, that means replacing `dependent.drv!out` by
+    `out(dynamic(dependent.drv))`
+  - Here again, we notice that `dynamic(transitivelyDependent.drv)` is the same as before,
+    so we don't build anything
+
+## Nix-build process
+
+### Aliases paths
+
+To allow this, we add a new type of store path: aliases paths.
+These paths don't actually exist in the store, just in the database and point to
+another path (so they are morally symlinks, but inside the db rather than
+on-disk)
+
+### Building a ca derivation
+
+ca derivations are derivations with the `__contentAddressed` argument set to
+`true`.
+
+The process for building a content-adressed derivation is the following:
+
+- We build it like a normal derivation to get an output path `$out`.
+- We compute a cryptographic hash `$chash` of `$out`[^modulo-hashing]
+- We move `$out` to `/nix/store/$chash-$name`
+- We create an alias path from `$out` to `/nix/store/$chash-$name`
+
+[^modulo-hashing]:
+
+  We can possibly normalize all the self-references before
+  computing the hash and rewrite them when moving the path to handle paths with
+  self-references, but this isn't strictly required for a first iteration
+
+### Building a normal derivation
+
+The process for building a normal derivation is the following:
+
+- We look into the drv for all the inputs paths of the build
+- For each input path, we look whether the path is an alias. If so we replace it
+  by its target
+- We compute the `dynamic` output of the derivation from the patched version
+- We then try to substitute and build the new derivation
+- We create an alias path from the `static` output to the `dynamic` one
+
+## Wrapping it up
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- Obviously, this makes the Nix model more complicated than what it is now. In
+  particular, the caching model needs some modifications (see [caching]);
+
+- We specify that only a sub-category of derivations can safely be marked as
+  `contentAddressed`, but there's no way to enforce these restricitions;
+
+- This will probably be a breaking-change for some tooling since the output path
+  that's stored in the `.drv` files doesn't correspond to the actual on-disk
+  path the output will be stored in (because it might just be an alias for the
+  other path)
+
+# Alternatives
+
+[alternatives]: #alternatives
+
+[RFC 0017][] is another proposal with the
+same end-goal. The big difference between these two is in the scope they cover:
+RFC 0017 is about fundamentally changing the base model of Nix, while this
+proposal suggests to make only the minimal amount of changes to the current
+model to allow the content-adressed model to live in parallel (which would open
+the way to a fully content-adressed store as RFC0017, but in a much more
+incremental way).
+
+Eventually this RFC should be subsumed by RFC0017.
+
+# Unresolved questions
+
+[unresolved]: #unresolved-questions
```

Comment:
Should we have functionality that allows to build a CA package twice with _different_ apparent output paths, and optionally with different parallelism settings? The build of the package obviously fails if the CA unification doesn't lead to the same result.

Should we mandate that Hydra uses this functionality? Should it be on by default?

Reply:
Per https://github.com/NixOS/rfcs/pull/62/files#r357243841 I think we can deal with non-deterministic derivation just fine.

Reply:
Well, for binary cache transparency it is much better if you can build something locally, then regain connectivity and fetch stuff from a cache, then fetch stuff from a different cache, then build some more locally, etc.

Reply:
So in my mind that's equally risky with and without content addressable derivations. The only difference is one lets you know if something goes wrong, and one doesn't.

Reply:
No. Build nondeterminism doesn't introduce significant behaviour changes, so as long as the expectations are not broken (yeah, we install you into this output path and your dependencies into those paths, and that is not going to change), it will be mostly usable. There are a few CPU-dependent optimisations from time to time, they are annoying.

With CA things are actually moved around, so even though everything would still work when assembled together, the assembling part will be failing. It is Nix, not the code that is built by Nix, that would fail to do things because of nondeterminism.

Reply:
I think trying to keep going despite non-determinism is a misfeature. You can always evict your own CA mappings (can keep the builds themselves for easy "rollback") and align with `cache.reflex-frp.org` and keep going.

Reply:
I think if things _are_ marked CA, of course it is a good idea to catch failures. But what you propose will not catch much, because a typical derivation is only built once (ever) by Hydra, later Hydra will use the binary cache. Also my proposal includes feeding different «apparent» output paths to the same build with the same dependencies, which has a better chance of discovering compressed self-references.

Reply:
I am OK with running `--check` and trying different termporary output paths. Catching non-determinism I don't think is important, because it's really *clashes* that we care about. However, catching self-references is important as we have to be able to move the thing.

Reply:
Well, varying the output paths is something that doesn't follow from anything Nix does, so it has to be spelled explicitly.